Add docs for GTM v4 (#1103)
snowplow/documentation: docs/sources/trackers/google-tag-manager/snowplow-template/tag-types/index.md

### Ad Tracking

The Ad Tracking tag is used to track impressions and ad clicks. This can used by, for example, ad networks to identify which sites and web pages users visit across a network, so that they can be segmented.

<details plain>
<summary>Ad Tracking Parameters</summary>

All ad tracking events take the following common parameters:

| Name           | Required? | Description                                                            | Example |
| -------------- | --------- | ---------------------------------------------------------------------- | ------- |
| `advertiserId` | No        | The advertiser ID                                                      | 201     |
| `campaignId`   | No        | The campaign ID                                                        | 12      |
| `cost`         | No        | The cost of the ad                                                     | 5.5     |
| `costModel`    | No        | The cost model for the campaign. Must be one of `cpc`, `cpm`, or `cpa` | cpc     |

The ad tracking tag includes three event types, each with its own set of additional parameters:

> #### Impression Event

Reviewer: level 4 headings is too small, it just looks like normal text

</details>

<details plain>
<summary>Enhanced Consent Parameters</summary>

| Attribute            | Description                                                                         | Example                 |
| -------------------- | ----------------------------------------------------------------------------------- | ----------------------- |
| `basisForProcessing` | GDPR lawful basis for data collection & processing.                                 | Consent                 |
| `consentUrl`         | URI of the privacy policy related document.                                         | https://example.com     |
| `consentVersion`     | Version of the privacy policy related document.                                     | 1.0                     |
| `consentScopes`      | The scopes allowed after the user finalized their selection of consent preferences. | 'analytics, functional' |
| `domainsApplied`     | The domains for which this consent allows these preferences to persist to.          | https://example.com     |
| `gdprApplies`        | Determine if GDPR applies based on the user's geo-location.                         | -                       |

</details>

> ### Enhanced Ecommerce

Reviewer: why Enhanced Ecommerce in twice? also make it clear that this one is sort-of deprecated

Author: I've combined them into one, and added a little note about being deprecated.

The Enhanced Ecommerce tag is used to track ecommerce events, such as product views, add to cart, and purchases.

See the [Enhanced Ecommerce Tag](/docs/sources/trackers/google-tag-manager/ecommerce-tag-template/index.md) template for more information.

### Error Tracking

The Errors tracker tag provides a way to manually track any errors that occur on your website.

<details plain>
<summary>Error Tracking Parameters</summary>

| **Name**   | **Required?** | **Description**                     | **Example**              |
| ---------- | ------------- | ----------------------------------- | ------------------------ |
| `message`  | Yes           | Error message                       | 'Cannot get user object' |
| `filename` | No            | Filename or URL                     | 'shop.js'                |
| `lineno`   | No            | Line number of problem code chunk   | 23                       |
| `colno`    | No            | Column number of problem code chunk | 45                       |
| `error`    | No            | JS `ErrorEvent`                     | {Stack trace string}     |

</details>

### Page View

> You can provide a Custom Page Title if you wish, and you can add a [custom context](/docs/sources/trackers/snowplow-tracker-protocol/custom-events/context-entities/index.md) to the request, as also described [here](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/page-views/index.md#tracking-a-page-view).

Reviewer: ```suggestion
You can provide a custom page title if you wish, and you can add a [custom context entity](/docs/sources/trackers/snowplow-tracker-protocol/custom-events/context-entities/index.md) to the request, as also described [here](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/page-views/index.md#tracking-a-page-view).
```

> If you enable Page Activity Tracking, page pings will be sent at intervals that you can specify.

Reviewer: what is Page Activity Tracking? link

Author: A field name, is formatted as code now.

> The Callback Function is something you can set to a JavaScript function. If you set the callback, then instead of sending the page ping to Snowplow, the function gets invoked instead.

Reviewer: what's with the random capitalisation? are these parameters? should be in a table if so like for the other sections on this page. Or at least marked up as code

is it definitely that the callback is instead of sending the ping, not as well?

Author: They're names of the fields in the UI, I'll format them as code.

And yeah, it's [this](https://docs.snowplow.io/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/activity-page-pings/#activity-tracking-callback) in the JS tracker, replaces the pings. I'll link it too.

### Link Click Tracking

The Track Link Click event is similar to regular parameter-based events, as it lets you add parameters and track a link click as a manually encoded hit.

The Enable Automatic Link Click Tracking adds listeners to the page, which will track clicks on links permitting they adhere to the blacklisted/whitelisted class names you can optionally provide. It is intended to trigger [early on a page](https://support.google.com/tagmanager/answer/7679319) to begin listening for click events, _not_ to trigger on click events like [Click Triggers](https://support.google.com/tagmanager/answer/7679320).

The Fix Middle-click Tracking adds a fix for some browsers where middle-clicks were not tracked properly.

If you check Track HTML Content Of Clicked Link, then the full text content of the link element will be sent to Snowplow as well.

:::note

This setting enables the tracker's native [Link Click Tracking plugin](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/link-click/index.md), it does not directly fire an event.

:::

### Form Tracking

> Form tracking has just two options, because filters and transformations won’t work with the custom template (due to lack of support for processing HTML elements). The options are to set form and/or field blacklists and whitelists.

Reviewer: are they really still called black/white lists or allow/deny lists?

Author: They are in fact allow/deny now, good catch!

Note that this tag enables the tracker's native [Form Tracking plugin](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/form-tracking/index.md), it does not directly fire an event.
It is intended to trigger [early on a page](https://support.google.com/tagmanager/answer/7679319) to begin listening for form events, _not_ to trigger on form events like [Form Submit Triggers](https://support.google.com/tagmanager/answer/7679217).

### Enhanced Ecommerce

When you select [Enhanced Ecommerce](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/ecommerce/enhanced/index.md), you are left with two options: Use Data Layer or Choose Variable. The way it works is very similar to Enhanced Ecommerce in Google Analytics.

If you choose the first, the template will look into the `dataLayer` structure for the most recently pushed [Enhanced Ecommerce object](https://www.simoahava.com/analytics/enhanced-ecommerce-guide-for-google-tag-manager/#data-types-actions), and map this object to the request to Snowplow Analytics.

If you selected Choose Variable, you need to provide a GTM variable that returns an object in the correct, expected format.

> For Google Analytics 4 compatible ecommerce tracking, see the dedicated [Snowplow v3 Ecommerce Tag](/docs/sources/trackers/google-tag-manager/previous-versions/v3/v3-tags/ecommerce-tag-template/index.md) template instead.

Reviewer: move this to the top of this subsection

> ### Custom Commands

Reviewer: ```suggestion
### Custom commands
```

> For any other commands which are supported by the Snowplow JavaScript Tracker v4, you can select the `[Custom Command]` option. Once selected, you can enter any function name and the associated parameter for that function. The parameter can either be a simple string, in examples such as `setUserId`:

Reviewer: where would I find a list of commands

![setUserId](images/setUserId.png)

Or it can be set to a Custom JavaScript Variable in the instances where an Object should be passed to the function, such as with `enableAnonymousTracking`:

![enableAnonymousTracking Custom JavaScript variable](images/enableAnonymousTracking_custom_variable.png)

And then use this variable as your Command Argument:

![enableAnonymousTracking Custom Command argument](images/enableAnonymousTracking_argument.png)

### Additional Tracking Parameters

![](images/additional_tracking_parameters.png)

> #### Add Custom Context Entities

Reviewer: ```suggestion
#### Add custom context entities
```

Using the Context Entities table allows you to attach [custom context entities](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/index.md#custom-context) to the Snowplow event. Each row should be set to a variable value that must be an array of custom context objects that will all be concatenated to add to the Event.

For example to manually attach the web page context, create a custom JavaScript variable that returns an array with the custom context object:

```js
function() {
    return [{
        schema: 'iglu:com.example/web_page/jsonschema/1-0-0',
        data: {
            id: '12345',
            title: 'Example Page',
            url: 'https://example.com'
        }
    }];
} 
```

Then set this variable as the value in the Context Entities table.

#### Set Custom Timestamp

> You can also choose to [set the True Timestamp](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/index.md#setting-the-true-timestamp) with this field. The format must be UNIX time in milliseconds.

Reviewer: ```suggestion
You can also choose to [set the True Timestamp](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/index.md#adding-custom-timestamps-to-events) with this field. The format must be UNIX time in milliseconds.
```

:::

## Form Tracking

Form tracking has just two options, because filters and transformations won’t work with the custom template (due to lack of support for processing HTML elements). The options are to set form and/or field denylists and allowlists.

Denylists and allowlists for forms are a list of comma-separated HTML class names. If denylisted, then any form element with a listed class will not trigger the form event. If allowlisted, then _only_ form elements with a listed classname will be tracked.

For fields, denylists and allowlists work similarly, except they use the `name` attribute rather than the class.

Note that this tag enables the tracker's native [Form Tracking plugin](/docs/sources/trackers/javascript-trackers/web-tracker/tracking-events/form-tracking/index.md), it does not directly fire an event.
It is intended to trigger [early on a page](https://support.google.com/tagmanager/answer/7679319) to begin listening for form events, _not_ to trigger on form events like [Form Submit Triggers](https://support.google.com/tagmanager/answer/7679217).

> ## Enhanced Ecommerce

Reviewer 2: What's the intent here?

Author: A header I didn't take out 😅
